tsdb/store.go

```diff
@@ -260,41 +270,48 @@ func (s *Store) CreateShard(database, retentionPolicy string, shardID uint64) er
 func (s *Store) DeleteShard(shardID uint64) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
+	return s.deleteShard(shardID)
+}
 
+// deleteShard removes a shard from disk. Callers of deleteShard need
+// to handle locks appropriately.
+func (s *Store) deleteShard(shardID uint64) error {
 	// ensure shard exists
-	sh, ok := s.shards[shardID]
-	if !ok {
+	sl, ok := s.shardLocations[shardID]
+	if !ok || sl.Shard == nil {
 		return nil
 	}
 
-	if err := sh.Close(); err != nil {
+	if err := sl.Shard.Close(); err != nil {
 		return err
 	}
 
-	if err := os.RemoveAll(sh.path); err != nil {
+	if err := os.RemoveAll(sl.Shard.path); err != nil {
 		return err
 	}
 
-	if err := os.RemoveAll(sh.walPath); err != nil {
+	if err := os.RemoveAll(sl.Shard.walPath); err != nil {
 		return err
 	}
 
-	delete(s.shards, shardID)
-
+	delete(s.shardLocations, shardID)
 	return nil
 }
 
-// DeleteDatabase will close all shards associated with a database and remove the directory and files from disk.
-func (s *Store) DeleteDatabase(name string, shardIDs []uint64) error {
+// DeleteDatabase will close all shards associated with a database and
+// remove the directory and files from disk.
+func (s *Store) DeleteDatabase(name string) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	for _, id := range shardIDs {
-		shard := s.shards[id]
-		if shard != nil {
-			shard.Close()
+	// Close and delete all shards on the database.
+	for shardID, location := range s.shardLocations {
+		if location.IsDatabase(name) {
+			// Delete the shard from disk.
+			if err := s.deleteShard(shardID); err != nil {
+				return err
+			}
 		}
-		delete(s.shards, id)
```

Comment:
Yes. Previously `DeleteDatabase` duplicated the logic of deleting a shard because `DeleteShard` fights for a lock on `s.mu`. So I moved the shard deleting logic out to `deleteShard` so `DeleteShard` and `DeleteDatabase` can share it.